numpy/lib/stride_tricks.py

```diff
@@ -111,6 +112,206 @@ def as_strided(x, shape=None, strides=None, subok=False, writeable=True):
     return view
 
 
+def _sliding_window_view_dispatcher(x, window_shape, axis=None, *,
+                                    subok=None, writeable=None):
+    return (x,)
+
+
+@array_function_dispatch(_sliding_window_view_dispatcher, module='numpy')
+def sliding_window_view(x, window_shape, axis=None, *,
+                        subok=False, writeable=False):
+    """
+    Create a sliding window view into the array with the given window shape.
+
+    Creates a sliding window view of the N dimensional array with the given
+    window shape. Window slides across each dimension of the array and extract
+    a subsets of the array at any window position.
+
+    Parameters
+    ----------
+    x : array_like
+        Array to create the sliding window view from.
+
+    window_shape : int or tuple of int
+        Size of window over each axis that takes part in the sliding window.
+        If `axis` is not present, must have same length as the number of input
+        array dimensions. Single integers `i` are treated as if they were the
+        tuple `(i,)`.
+
+    axis : int or tuple of int, optional
+        Axis or axes along which the sliding window is applied.
+        By default, the sliding window is applied to all axes and
+        `window_shape[i]` will refer to axis `i` of `x`.
+        If `axis` is given as a `tuple of int`, `window_shape[i]` will refer to
+        the axis `axis[i]` of `x`.
+        Single integers `i` are treated as if they were the tuple `(i,)`.
+
+    subok : bool, optional
+        If True, sub-classes will be passed-through, otherwise the returned
+        array will be forced to be a base-class array (default).
+
+    writeable : bool, optional
+        When true, allow writing to the returned view. The default is false,
+        as this should be used with caution: the returned view contains the
+        same memory location multiple times, so writing to one location will
+        cause others to change.
+
+    Returns
+    -------
+    view : ndarray
+        Sliding window view of the array. The sliding window dimensions are
+        inserted at the end, and the original dimensions are trimmed as
+        required by the size of the sliding window.
+
+        That is, `view.shape = x_shape_trimmed + window_shape`, where
+        `x_shape_trimmed` is `x.shape` with every entry reduced as follows.
+
+        Let `input_size` be the size of the input array in an axis
+        participating in the sliding window, and let `window_size` be the
+        window size in the corresponding window axis, then the respective entry
+        of `x_shape_trimmed` will be `input_size - (window_size - 1)`.
```

Comment:
This is a little wordy now :).

Maybe we could just give the formula, i.e.,
```suggestion
        Sliding window view of the array. The sliding window dimensions are
        inserted at the end, and the original dimensions are trimmed as
        required by the size of the sliding window. That is,
        `view.shape = x_shape_trimmed + window_shape` where
        `x_shape_trimmed[i] = x.shape[i] - (window_shape[i] - 1)`.
```

Reply:
I completely agree that it is more wordy than desirable. The problem is that your suggested formula is not correct and while I was trying to come up with a correct formula, I ended up with things that I did not find more enlightening than the text. The correct formula would be `x_shape_trimmed[i] = x.shape[i] if i not in axis` and otherwise
```
x_shape_trimmed[axis[j]] = x.shape[axis[j]] - (window_shape[j] - 1)
```
except for the doubly used axis, in which case you need to subtract the contribution from all window axis (avoid double counting).

Reply:
Right - the challenge here is to describe exactly what these lines of code do:

https://github.com/numpy/numpy/blob/97315fc86e7e8dc49aa659eb86428c915bf02375/numpy/lib/stride_tricks.py#L304-L311

IMO we maybe want to lean more in a "they do the obvious thing" direction - what's important is that:
* The new `window_shape` dimensions are added at the end
* The number of initial dimensions are unchanged, but the length of each dimensions will be adjusted to account for the windowing